dsp/modules/pyserini.py

```diff
@@ -0,0 +1,73 @@
+from typing import Union
+import json
+from datasets import Dataset
+
+from dsp.utils import dotdict
+from pyserini.search import FaissSearcher
+from pyserini.prebuilt_index_info import TF_INDEX_INFO, FAISS_INDEX_INFO, IMPACT_INDEX_INFO
+
+
+class PyseriniRetriever:
+    """Wrapper for retrieval with Pyserini. Supports using either pyserini prebuilt faiss indexes or your own faiss index."""
+
+    def __init__(self, 
+                 query_encoder: str = 'castorini/dkrr-dpr-nq-retriever', 
+                 index: str = 'wikipedia-dpr-dkrr-nq', 
+                 dataset: Dataset = None,
+                 id_field: str = 'id',
+                 text_fields: list[str] = ['text']) -> None:
+        """
+        Args:
+        
+            query_encoder (`str`):
+                Huggingface model to encode queries
+            index (`str`):
+                Either a prebuilt index from pyserini or a local path to a faiss index
+            dataset (`Dataset`):
+                Only required when using a local faiss index. The dataset should be the one that has been put into the faiss index.
+            id_field (`str`):
+                The name of the id field of the dataset used for retrieval.
+            text_fields (`list[str]`):
+                A list of the names of the text fields for the dataset used for retrieval.
+        """
+        
+        self.encoder = FaissSearcher._init_encoder_from_str(query_encoder)
+        self.dataset = dataset
+        self.id_field = id_field
+        self.text_fields = text_fields
+        
+        if index in TF_INDEX_INFO or index in FAISS_INDEX_INFO or index in IMPACT_INDEX_INFO:
+            self.searcher = FaissSearcher.from_prebuilt_index(index, self.encoder)
+        else:
+            self.searcher = FaissSearcher(index_dir=index, query_encoder=self.encoder)
+
+    def __call__(
+        self, query: str, k: int = 10, simplify: bool = False, threads: int = 16,
+    ) -> Union[list[str], list[dotdict]]:
+        hits = self.searcher.search(query, k=k, threads=threads)
+        
+        topk = []
+        for rank, hit in enumerate(hits, start=1):
+            if self.dataset is not None:
+                # Search through user's dataset for the document (could be slow), maybe room for improvement?
```

Comment:
what would make it slow? isn't it just a lookup? or does it go to disk?

Reply:
yeah it's just a linear lookup through the dataset

Reply:
Yeah some room for the future improvements!

Reply:
for the indexible Dataset, it'll be faster to create a map from docid to index during initialization, so maybe we can implement that? @okhat if you have any other ideas let me know as well

Reply:
It’s a linear lookup? Or constant time lookup?